cdap-app-templates/cdap-program-report/cdap-program-report1_2.10/src/main/scala/co/cask/cdap/report/ReportAggregationFunction.scala

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.report
+
+import java.util.concurrent.TimeUnit
+
+import co.cask.cdap.api.schedule.{TriggerInfo, TriggeringScheduleInfo}
+import co.cask.cdap.report.util.{Constants, TriggeringScheduleInfoAdapter}
+import com.google.gson.GsonBuilder
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
+import org.apache.spark.sql.expressions.{MutableAggregationBuffer, UserDefinedAggregateFunction}
+import org.apache.spark.sql.types._
+import org.slf4j.LoggerFactory
+
+/**
+  * An aggregation function that aggregates [[Row]]'s with the same program run ID into intermediate
+  * [[MutableAggregationBuffer]]'s and finally merge the [[MutableAggregationBuffer]]'s to build a single [[Row]]
+  */
+class ReportAggregationFunction extends UserDefinedAggregateFunction {
+  import ReportAggregationFunction._
+
+  override def inputSchema: StructType = new StructType()
+    .add(Constants.NAMESPACE, StringType, false)
+    .add(Constants.APPLICATION_NAME, StringType, false)
+    .add(Constants.APPLICATION_VERSION, StringType, false)
+    .add(Constants.PROGRAM_TYPE, StringType, false)
+    .add(Constants.PROGRAM, StringType, false)
+    .add(Constants.RUN, StringType, false)
+    .add(Constants.STATUS, StringType, false)
+    .add(Constants.TIME, LongType, false)
+    .add(Constants.MESSAGE_ID, StringType, false)
+    .add(Constants.START_INFO, new StructType()
+      .add(Constants.USER, StringType, true)
+      .add(Constants.RUNTIME_ARGUMENTS, MapType(StringType, StringType), false)
+      .add(Constants.ARTIFACT_ID, new StructType()
+        .add(Constants.ARTIFACT_NAME, StringType, false)
+        .add(Constants.ARTIFACT_SCOPE, StringType, false)
+        .add(Constants.ARTIFACT_VERSION, StringType, false), false),
+      true)
+
+  override def bufferSchema: StructType = new StructType()
```

Comment:
rename it as `output/record Schema` or something else. `bufferSchema` is generic.